### DISCUSSION TITLE
Proposal: Split async methods into separate interfaces?

### DISCUSSION THREAD

**Discussion by JesseTG** [dre0dru/LocalStorage#4] [Ideas]
Some storage backends only support synchronous APIs. Some only support *asynchronous* APIs. Would it be possible to divide interfaces that have methods for both into synchronous and asynchronous versions?

**Comment by dre0dru** [comment]
@JesseTG Hello! Can you please provide more info on how you view this? As I understand you are talking about `IFileStorage` and `IPlayerPrefsStorage` and they support async and sync API which is defined mostly by underlying `ISerializationProvider`. Default package implementation (`UnityJsonSerializationProvider`) indeed provides only sync API for serialization, but some serialization libraries like [UT8Json](https://github.com/neuecc/Utf8Json) provide async API, too. Also `IDataTransofrm` implementations that can be used with `ISerializationProvider` also have async API.

Right now my only thoughs are something like this:
```c#
interface IFileStorage {
//sync API + utility methods
}

interface IAsyncFileStorage : IFileStorage {
//addintional async API
}

class FileStorage : IAsyncFileStorage {
// both async and sync implementation
}
```
The same for `IPlayerPrefsStorage`.
This way, e.g., you can bind instance to sync or async interface.
Maybe you can provide use case where you need that kind of separation if I misunderstood you?

**Comment by JesseTG** [comment]
No, you understand me correctly.

I don't have a _specific_ use case right now, since I happen to almost exclusively use asynchronous storage APIs in my project, but I'm trying to think a step ahead. Maybe just keep this idea in your back pocket? Honestly, if you had discussions enabled I'd have mentioned this there instead of opening a ticket.

**Reply by dre0dru** [reply]
Moved to discussions then.
To summarise, you want to have ability to choose between 3 set of APIs for storage interfaces:
1. Sync
2. Async
3. Async + Sync

I will keep it in mind and think about it.

**Reply by JesseTG** [reply]
Pretty much. Since as it exists right now, new implementations would have to support both interfaces, which may not always be possible.

Although item 3 would just be items 1 and 2, so really it would be two sets of APIs.

**Reply by dre0dru** [reply]
@JesseTG Created preliminary [PR](https://github.com/dre0dru/LocalStorage/pull/5) for this feature. Can you check it?
Basically, all interfaces are split into two:

```
interface IInterface
{
//sync API
//core utility methods if present
}

interface IInterfaceAsync : IInterface
{
//async API
}
```
This way you can use either sync or sync+async API depending on what iterface your implementation depends on.

I am still thinking on what to do with implementations, that doesn't really support async API. E.g. `UnityJsonSerializationProvider`. Right now async API implemented as fake async by returning completed `Task` with result from sync API. Unity doesn't have an async json serialization API, and I don't really want to put tons of different implementations in package to support all combinations of async/fake async/sync dependencies, so implementations are subject to change.

Btw, do you have any examples where that kind of sync/async API separation implemented? I really like that idea, just don't know how to cleanly implement it with minimal amount of classes in package.

**Reply by JesseTG** [reply]
> 
> 
> @JesseTG Created preliminary [PR](https://github.com/dre0dru/LocalStorage/pull/5) for this feature. Can you check it?
> Basically, all interfaces are split into two:
> 
> ```
> interface IInterface
> {
> //sync API
> //core utility methods if present
> }
> 
> interface IInterfaceAsync : IInterface
> {
> //async API
> }
> ```

Some cloud storage APIs only work asynchronously, so I'd go with something like this:

```csharp
interface IInterfaceSync
{
}

interface IInterfaceAsync
{
}

interface IInterface : IInterfaceSync, IInterfaceAsync
{
}
```

This way compatibility is a compile-time issue; if a cloud storage service doesn't provide synchronous APIs, an implementation can simply not use `IInterfaceSync` instead of throwing `NotSupportedException`s all over the place.

Put another way: if .NET provides [`IReadOnlyCollection<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ireadonlycollection-1?view=netstandard-2.0), then why does [`ICollection<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.icollection-1?view=netstandard-2.0) provide a [`IsReadOnly`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.icollection-1.isreadonly?view=netstandard-2.0#System_Collections_Generic_ICollection_1_IsReadOnly) property? If that property needs to be true, then why not just implement `IReadOnlyCollection<T>` without also implementing `ICollection<T>`?

Both `IInterfaceSync` and `IInterfaceAsync` can define the same set of common methods, and a class that defines both can just implement them once (unless they'd like to implement both with [explicit implementations](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/interfaces/explicit-interface-implementation)).

> I am still thinking on what to do with implementations, that doesn't really support async API. E.g. `UnityJsonSerializationProvider`. Right now async API implemented as fake async by returning completed `Task` with result from sync API. Unity doesn't have an async json serialization API, and I don't really want to put tons of different implementations in package to support all combinations of async/fake async/sync dependencies, so implementations are subject to change.

That's okay, just implement the synchronous APIs and provide extension methods to make them async. Something like this:

```csharp
public static class SerializationExtensions
{
    public static UniTask<byte[]> SerializeAsync<T>(this ISerializationProviderSync provider, T data)
    {
        return UniTask.FromResult(/* etc */);
    }
}
```

If an extension method and an instance method with the same signature exist, [the instance method will be prioritized.](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/how-to-implement-and-call-a-custom-extension-method#net-security)

> Btw, do you have any examples where that kind of sync/async API separation implemented? I really like that idea, just don't know how to cleanly implement it with minimal amount of classes in package.

Sort of. [MessagePipe](https://github.com/Cysharp/MessagePipe) comes to mind (see [this section](https://github.com/Cysharp/MessagePipe#publishsubscribe)), but it doesn't provide interfaces that offer both sync and async APIs.

**Reply by dre0dru** [reply]
Now I managed to [implement](https://github.com/dre0dru/LocalStorage/pull/7) it right I think 😁
New version is out, some namespaces are changed, so expect compile errors on update.

**Reply by JesseTG** [reply]
Thank you, I'll give it a try tomorrow and let you know how it goes!
